File: sdk/typescript/src/streaming.ts
/** SSE streaming handler for Samhati SDK. */

import type { ChatCompletionChunk } from "./types.js";

/**
 * Parse an SSE stream from a fetch Response into an async generator
 * of ChatCompletionChunk objects.
 */
export async function* parseSSEStream(
  response: Response
): AsyncGenerator<ChatCompletionChunk> {
  const reader = response.body?.getReader();
  if (!reader) {
    throw new Error("Response body is null");
  }

  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Process complete lines
      const lines = buffer.split("\n");
      // Keep the last (possibly incomplete) line in the buffer
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || trimmed.startsWith(":")) continue;
        if (!trimmed.startsWith("data: ")) continue;

        const payload = trimmed.slice("data: ".length).trim();
        if (payload === "[DONE]") return;

        const chunk: ChatCompletionChunk = JSON.parse(payload);
        yield chunk;
      }
    }
  } finally {
    reader.releaseLock();
  }
}

File: sdk/rust/src/streaming.rs
//! SSE streaming handler for the Samhati SDK.

use futures::stream::Stream;
use reqwest::Response;
use std::pin::Pin;
use std::task::{Context, Poll};

use crate::error::{Result, SamhatiError};
use crate::types::ChatCompletionChunk;

/// A stream of `ChatCompletionChunk` parsed from an SSE response.
pub struct SseStream {
    inner: Pin<Box<dyn Stream<Item = Result<ChatCompletionChunk>> + Send>>,
}

impl SseStream {
    pub fn new(response: Response) -> Self {
        let stream = async_stream::stream! {
            let mut buffer = String::new();
            let mut bytes_stream = response.bytes_stream();

            use futures::StreamExt;
            while let Some(result) = bytes_stream.next().await {
                let bytes = match result {
                    Ok(b) => b,
                    Err(e) => {
                        yield Err(SamhatiError::Http(e));
                        return;
                    }
                };

                let text = match std::str::from_utf8(&bytes) {
                    Ok(t) => t.to_string(),
                    Err(e) => {
                        yield Err(SamhatiError::SseParse(e.to_string()));
                        return;
                    }
                };

                buffer.push_str(&text);

                while let Some(newline_pos) = buffer.find('\n') {
                    let line = buffer[..newline_pos].to_string();
                    buffer = buffer[newline_pos + 1..].to_string();

                    let trimmed = line.trim();
                    if trimmed.is_empty() || trimmed.starts_with(':') {
                        continue;
                    }

                    if let Some(payload) = trimmed.strip_prefix("data: ") {
                        let payload = payload.trim();
                        if payload == "[DONE]" {
                            return;
                        }

                        match serde_json::from_str::<ChatCompletionChunk>(payload) {
                            Ok(chunk) => yield Ok(chunk),
                            Err(e) => {
                                yield Err(SamhatiError::SseParse(format!(
                                    "Failed to parse chunk: {e}"
                                )));
                                return;
                            }
                        }
                    }
                }
            }
        };

        Self {
            inner: Box::pin(stream),
        }
    }
}

impl Stream for SseStream {
    type Item = Result<ChatCompletionChunk>;

    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        self.inner.as_mut().poll_next(cx)
    }
}
